File: vake/tty.py
from enum import Enum
from typing import List


class Color(Enum):
    BLACK = 30
    RED = 31
    GREEN = 32
    YELLOW = 33
    BLUE = 34
    MAGENTA = 35
    CYAN = 36
    WHITE = 37
    RESET = 39

    def decorate(self, obj: object, bold: bool = False) -> str:
        base = f'{self.escape()}{obj}{Color.RESET.escape()}'
        if bold:
            return f'\033[1m{base}\033[0m'
        else:
            return base

    def escape(self):
        return f'\033[{self.value}m'


class Looper:
    __seq: List[str]
    __i: int

    @classmethod
    def spin(cls, clockwise: bool = True) -> 'Looper':
        seq = ['|', '/', '-', '\\']
        if clockwise:
            return Looper(seq)
        else:
            return Looper(list(reversed(seq)))

    @classmethod
    def dots(cls, clockwise: bool = True) -> 'Looper':
        seq = ['⣷', '⣯', '⣟', '⡿', '⢿', '⣻', '⣽', '⣾']
        if clockwise:
            return Looper(seq)
        else:
            return Looper(list(reversed(seq)))

    def __init__(self, seq: List[str]):
        self.__seq = seq
        self.__i = 0

    def __next__(self):
        char = self.__seq[self.__i]
        self.__i = (self.__i + 1) % len(self.__seq)
        return char

    def __iter__(self):
        return self

File: app/tty.ts
export class Color {
  static readonly BLACK = new Color(30);
  static readonly RED = new Color(31);
  static readonly GREEN = new Color(32);
  static readonly YELLOW = new Color(33);
  static readonly BLUE = new Color(34);
  static readonly MAGENTA = new Color(35);
  static readonly CYAN = new Color(36);
  static readonly WHITE = new Color(37);
  static readonly RESET = new Color(39);

  readonly code: number;

  private constructor(code: number) {
    this.code = code;
  }

  apply(
    message: string,
    options: { bold?: boolean } = {},
  ): string {
    const base = `${this}${message}${Color.RESET}`;
    return options.bold ? `\x1b[1m${base}\x1b[0m` : base;
  }

  toString(): string {
    return `\x1b[${this.code}m`;
  }
}
